### PR TITLE
Feature/tesv1.1

### PR DESCRIPTION
**Pull request by lbeckman314**
Adds tests/checks for the following `v1.1` features:

- #15
- #16
- #17
- #18
- #19
- #20
- #21
- #22

Happy to split into separate PR's if helpful!

### CODE REVIEW

**Comment by LakiG** [comment]
Hey Liam, thanks for the new tests. You can also create a branch in the repo and make a PR directly going forward.

File: tests/create_task_backend_parameters.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        "description": "CompTest",
+        "executors": [
+          {
+            "image": "ubuntu:20.04",
+            "command": [
+              "TestString1",
+              "TestString2"
+            ]
+          },
+          {
```
Comment (LakiG) [inline tests/create_task_backend_parameters.yml:28]: We can use a single executor here. Same for the other tests as well, if multiple executors are not being used functionally.

File: tests/create_task_backend_parameters.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+              "TestString3",
+              "TestString4"
+            ]
+          }
+        ],
+        "resources": {
+          "backend_parameters": {
+            "VmSize" : "Standard_D64_v3"
+          },
+          "backend_parameters_strict": false
```
Comment (LakiG) [inline tests/create_task_backend_parameters.yml:40]: If we keep the flag as `false`, this test would serve only as a schema check. For a feature test, it might be better to put the `backend_parameters_strict` as `true` and intentionally enter wrong values in the `backend_parameters`. The expected response code would be 4xx/5xx, but its not mentioned in the schema. Lets keep response code as `400` for now.

File: tests/create_task_ignore_error.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            "command": [
+              "error"
+            ],
+            "ignore_error": true
+          },
+          {
+            "image": "ubuntu:20.04",
+            "command": [
+              "TestString",
+              "TestString4"
```
Comment (LakiG) [inline tests/create_task_ignore_error.yml:32]: These commands will give an error. If the second command also fails, the overall task state will be `EXECUTOR_ERROR`, having the same outcome if  `ignore_error` was absent. We can add a working command instead and add a `GetTask` job to verify that the task state was `COMPLETE`.

File: tests/create_task_backend_parameters.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
+name: Create Tes Task Job With Backend Parameters
+description: Job to create a new TES Task with backend parameters
+service: TES
+compatible_versions:
+  - v1.1
+tags:
+  - logical
+  - create task
+  - create task test
```
Comment (LakiG) [inline tests/create_task_backend_parameters.yml:9]: Add a name which is specific only to this test. All create task tests are covered by 2nd tag. Same for the other tests as well.

File: tests/create_task_optional_filetype.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+description: Job to create a new TES Task intentionally without input and output filetypes
+service: TES
+compatible_versions:
+  - v1.1
+tags:
+  - logical
+  - create task
+  - create task test
+  - all
+  - dev
```
Comment (LakiG) [inline tests/create_task_optional_filetype.yml:11]: We can remove this tag perhaps.

File: tests/create_task_output.yml [hunk 6]
```diff
@@ -0,0 +1,47 @@
+name: Create Tes Task Job With Output
+description: Job to create a new TES Task with output
```
Comment (LakiG) [inline tests/create_task_output.yml:2]: Which new feature/schema is this test being added for?

File: tests/create_task_resources.yml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+              "TestString4"
+            ]
+          }
+        ],
+        "resources": {
+          "cpu_cores": 4,
+          "preemptible": false,
+          "ram_gb": 8,
+          "disk_gb": 40,
+          "zones": [
```
Comment (LakiG) [inline tests/create_task_resources.yml:42]: This test is also not related to the new `v1.1` features, right?  As for a general schema check, `resources` and its children are optional, so adding a test wont be beneficial.

File: tests/create_task_output_wildcard.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        ],
+        "outputs": [
+          {
+            "url": "https://my.storage.org/bucket/",
+            "path": "/path/in/container/*/*.?am",
+            "path_prefix": "/path/in/container"
+          },
+          {
+            "url": "https://my.storage.org/bucket/wildc*ard_char?",
+            "path": "/path/with/wildc\*rd_char\?"",
```
Comment (LakiG) [inline tests/create_task_output_wildcard.yml:44]: This test will fail earlier due to an invalid storage URL. Cloud-integrated tests are not supported currently. A proper test would include accessing a storage URL, inputting a file, defining the commands for output files and then verifying the above output path conditions. The issue is already raised for this and maybe a test can be added in future.

File: tests/create_task_streamable.yml [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            ]
+          }
+        ],
+        "input": [
+          {
+            "name": "access-file",
+            "description": "Access file resource via streaming",
+            "url": "s3://my-object-store/file1",
+            "path": "/data/file1",
+            "streamable": true
```
Comment (LakiG) [inline tests/create_task_streamable.yml:42]: The `optional_filetype` and `streamable` tests are just schema checks specifically for TES `v1.1`. Can we combine both tests? It will check both schemas in a single task.

**Review by LakiG — CHANGES_REQUESTED** [review]

**Comment by LakiG** [comment]
Sorry, I closed the PR due to a faulty merge. Reopening the PR now.

**Comment by uniqueg** [comment]
This is strange, I didn't delete any branch 3 minutes ago.
But I guess the target branch of the PR is deleted now because it was merged yesterday.

Can you rebase to `dev` please, @lbeckman314?
Sorry about that.

**Comment by uniqueg** [comment]
@lbeckman314: Anything I can do on here right now? It still says closed, I guess because the target branch doesn't exist anymore. Do you wanna change that? Sorry for the mess on our side, @lbeckman314

**Comment by lbeckman314** [comment]
No worries @uniqueg! I was rebasing the tests onto @LakiG's [feature/tes-v1_1_0](https://github.com/elixir-cloud-aai/tes-compliance-suite/tree/feature/tes-v1_1_0) branch. I will create a new PR to carry over these changes onto that branch if that works!

File: tests/create_task_backend_parameters.yml (hunk 4 above)
Comment (uniqueg) [reply tests/create_task_backend_parameters.yml:9]: Couldn't we use the task name as a test-specific "tag"? Makes no sense to add a(nother) test-specific to the tests...

Speaking of which (but a bit off topic): what happens if two tests with the same name are registered with the runner? Is there a defined behavior?

File: tests/create_task_resources.yml (hunk 7 above)
Comment (uniqueg) [reply tests/create_task_resources.yml:42]: Good point! If (and only if!) all properties (including all nested ones) are optional _and_ `additionalProperties` are not disabled on the schema or any child schemas, then tests like these don't add value.

And I think that's indeed the case here, and possibly elsewhere.

File: tests/create_task_output.yml (hunk 6 above)
Comment (uniqueg) [reply tests/create_task_output.yml:2]: I didn't check the other tests, but if there's no test with outputs, and if outputs - if available - has required properties (which I think it has), then I think such a test (with only the required output properties) makes sense.

See discussion in next comment for details.

File: tests/create_task_streamable.yml (hunk 9 above)
Comment (uniqueg) [reply tests/create_task_streamable.yml:42]: I think there is value in keeping them separate. It's more explicit, and if the behavior of one of these changes in the future, it will be clear where the problem is.

File: tests/create_task_backend_parameters.yml (hunk 4 above)
Comment (LakiG) [reply tests/create_task_backend_parameters.yml:9]: I followed this structure for tags. 1st -> Logical/Functional. 2nd -> The primary endpoint dependent on the test. 3rd -> The individual test-specific name, in order to execute a single test. 4-> All. 

Currently, all tests have a unique test-specific tag. So, using it will only run that particular test as intended. However, if an "endpoint" is entered as tag. All tests related to that endpoint are run.

File: tests/create_task_output.yml (hunk 6 above)
Comment (LakiG) [reply tests/create_task_output.yml:2]: Yes, you are right. Currently, there are no tests to check the outputs specifically.
Keeping v1.0.0 in mind, we can add a test having all required properties (3 in number). For v1.1.0, there are 2 required properties. I will log an issue for it and add the new tests.

Just wanted to discuss, how should we handle the test file names? Both are exactly same, just the version is different. Perhaps, for now we can keep the version in the file name. And when we separate the tests and executors, we can have test folders specific to versions.

File: tests/create_task_streamable.yml (hunk 9 above)
Comment (LakiG) [reply tests/create_task_streamable.yml:42]: Sure, for better visibility we can keep them separate. But if there is any change in behavior, we will need to create another test for that newer version as it wont be supported by v1.1.0.

File: tests/create_task_output.yml (hunk 6 above)
Comment (uniqueg) [reply tests/create_task_output.yml:2]: In that case, yes, please add a test for outputs for 1.0.0. And if there are other such cases (i.e., non-required properties that do have required child properties, if provided), then please add tests for these, too, if not already available.

As for the names: I think you should try and disentangle two roles here, that of the developer of the test runner and that of the developer of _test cases_ (now you do both, but the idea is that sooner or later, the second role will be done by other people, both for TES and maybe other specs). And I think that as a result of the confusion of these two roles, we are currently using too many different things for the same purpose (e.g., we have the test file path/name, the `name` field, and we put a unique tag for each test case).

What I would do, is to actually remove the `name` field altogether. It is difficult to come up with naming conventions that account for all possible cases, it is very difficult to enforce naming conventions, and you make it difficult for yourself either enforcing unique names (extra code, and then what? Erroring out?) or _not_ enforcing unique names (how do you report the results of a given test if you can't uniquely map it back). And last but not least, you put an unnecessary cognitive burden on the people writing tests. And for what benefit? That people hopefully use semantic names for tests? But they will still write a description (hopefully) and ... here's the thing ... give their test file a name - which, together with its path, is _guaranteed_ to be unique (note that you could also use a single path/filename as an argument rather than a directory of tests - this is how, e.g., `pytest` works as well; so no need for unique tags for a given test)!

So, I would say, for the runner, don't worry about the naming or the file subdirectories in which test developers organize the tests: Instead, get all the valid tests in the target directory and subdirectories, filter by tags, and use the filename with path as their unique identifiers. Users will then need to dig into the tests (perhaps a `--verbose` option would also print out the test description in the report? If it does not already do so), but what can you do? The test runner can't ever be generic _and_ understand the semantics of a set of tests at the same time. To make it easy for users (people running the tests) to understand what a given test does by describing it will in the field that is dedicated for it.

Likewise, with the versions: _Every_ test is specific to a set of versions. Anything in the API can change, at any time, so a test that _seems_ to be "generic" now, may not be so in TES 2.0.0, or TES 7.3.1. I see two ways of how to deal with it:
1. You treat versions as nothing special and tell test developers that they must use tags to support versioned APIs. And then don't have any version-specific code in your runner. Of course this option puts a bit of a burden on test developers if there are a lot of versions (who then need to enumerate them in the tags of each test) and also makes the tests more messy.
2. You _do_ recognize that versions are basically integral to API tests and build in proper support for them. In that case, to minimize code/test repetition, I would ask test developers to provide a mapping of tests (by path/filename) for each version. Then you can remove the version tags. And by default you can always use the mapping for the _lastest_ version (by SemVer ordering rules). This is what I would do, although option (1) seems fair to me as well.

Speaking of tags (and now moving to the test developer role), I find the `logical` and `functional` tags to be a bit ambiguous / not self evident. And again, I think that there are some integral things going on again that you could consider adding proper support for (outside of tags).

Basically, there are two types of tests, with the first type having itself two subtypes:
1. The service does what it is supposed to do. These are currently tagges as `functional`. Out of these, some may be (1a) required by the specification, while others are (1b) optional.
2. The service accepts what it is supposed to accept and returns what it is supposed to return. These tests are currently tagged as `logical`. I may be wrong, but I think all of these are, by definition, required by the specification.

In the end, we would probably be able to give a service a sort of "green badge" if - and only if! - all tests (for the correct schema version) that fill into categories (1a) and (2) succeed. Then, for optional functionalities, we would need to parse the report and see if a given service supports it or not.

There are some ways how to implement that:
1. Add two boolean flags: `required` and `functional`. The problem here is that optional non-functional tests probably don't exist, and also this setup would not be very extensible. `required` is probably truly boolean, it either is or isn't. But whether something is a functional test, or logical/validation or maybe something else is not so clear, as `functional` is not a true antonym of `logical`, and `non-functional` is a weird and unhelpful label. So maybe let's not do that.
2. Enumerate `validation_only`, `functional_required` and `functional_optional`. That would probably work and is extensible, but is difficult to parse, especially if it _is_ extended in the future.
3. So perhaps from the test runner side we don't really care if a test is `logical` or `functional` to give that green badge. We don't need to know whatever a test developer means when they label a test like that. It may mean something to _them_ and hopefully also to the user, but not to the test runner. So why not add a boolean flag `required` to the test schema, and make use of that to decide whether a test run passed or failed. And for further classifications, we still have the `tag` field.

For example, for our situtation, I would simply use a single tag `schema_validation_only` (more descriptive than `logical`). I don't really see the need/use for any other labels than that. If a user really _only_ wanted to run "functional" tests (which is a bit weird, but okay), and we wanted to support that, we could implement a negative tag filter for the runner.

Finally, I would only allow letters, digits and underscores in tag names; it's not much of a restriction for test developers, but it makes parsing from the command line much easier. If you want to filter by multiple tags, being able to do `--include tag1,tag2` is much more user-friendly than having to deal with quotes and backslash escpates and whatnot.

Okay, that became a very very long comment. Wasn't intended, so sorry about that.

## tl;dr

1. Remove the `name` field from test case schema and tags for individual test cases. Use the path/filename instead (already unique!). Allow users to specify paths to one or more test cases or directories of tests and run all tests that are found, filtered by tags, if available.
2. Remove version tags. Bake in support for versioning by asking test developers to provide mappings from versions to test case paths for each version. Add a functionality to discover available versions for the set of test cases (by parsing that dictionary). Use the most recent version as default if not version is supplied.
3. Provide `required` field in test case schema to decide if a test run passed or failed.
4. Give the user the ability to do a negative selection on tags (`--include-tags` vs `--exclude_tags`). Remove excessive tags from the available test cases, e.g., keep only one of `logical` (better: `schema_validation_only`) or `functional` tags. With positive and negative tag filtering, this will be powerful enough.

Those are my suggestions anyway. If you agree with any of these, please create issues for those, otherwise ignore :)

File: tests/create_task_backend_parameters.yml (hunk 4 above)
Comment (uniqueg) [reply tests/create_task_backend_parameters.yml:9]: I think you may be conflating your roles as test case developer and test runner developer a bit, leading to a bit of an abuse of tags, perhaps.

Tags are generally arbitrary key-values, it's difficult to enforce rules on those, and therefore it's dangerous to put any semantic assumptions on them. Things may work _now_ because you wrote both the runner code and the test cases and know all about them, but latest when someone else wants to use the runner for a different set of tests that you are not involved in, this will cause problems, for them and for you. See the next (very long) comment for a more detailed description of the problem(s) (it's a bit of a mess, sorry) and some suggestions on how to clean that up a bit.

File: tests/create_task_streamable.yml (hunk 9 above)
Comment (uniqueg) [reply tests/create_task_streamable.yml:42]: And I think that's totally fine. _Every_ test case is tied only to a specific set of versions, even if it isn't obvious to us now, or -for some tests- possibly never (see comment above).

But assuming we have figured out a way how to run the right tests for a given version, running a test run and having separate tests for `streamable` and the optional file type is convenient for the user: they will not have to try to disentangle whether a failed test is because of the one or the other or both.

File: tests/create_task_output.yml (hunk 6 above)
Comment (LakiG) [reply tests/create_task_output.yml:2]: Thank you @uniqueg for this detailed suggestion. I understood that these things might convolute the code/logic in future and it is best to simplify them as you mentioned.

For names: Mapping the file paths/names as unique identifiers will indeed serve as an easier alternative to defining unique names as done currently in every test.
We can add the test description in the report without the `--verbose` flag. They should not be lengthy and we will always be able to understand the purpose of the test.

In versions, I understand that with multiple versions, the test developers would need to add new versions in every test. But, the test developers are closer to the service and they should be able to manage this. 
Adding support on the runner side will complicate the division of responsibilities. For every test-version mapping change in the "service" API, the change would be on the "runner" (unrelated to any runner change). This will also create a direct dependency on the runner team.

In every way, we have to store the test to version mapping. The test developers (who deal with both, tests and versions) will be the better choice to store them than the test runners (who only deal with versions). We can maybe add this in future if we think that the version tags are becoming unmanageable.

Thinking about the optional `functional` tests, all the tests are based on the API specification. If a test is written, it should _pass_ for _all_ the API services. For an optional test, it suggests that it was _not_ defined in the specification, for which we don't need to write that test. I think all tests are required, irrespective of being `logical` or `functional.`

The negative tag filter is a great idea. The single tag `- schema_validation_only` can be efficiently used then.

The CLI arg for tag can be upgraded as you suggested. Adding constraints on the tags would allow this implementation.

**Comment by lbeckman314** [comment]
See #32 for updated PR based on this one.
